dydx/dydxPresenters/dydxPresenters/_v4/Analytics/TransferAnalytics.swift

```diff
@@ -0,0 +1,54 @@
+//
+//  TransferAnalytics.swift
+//  dydxPresenters
+//
+//  Created by Rui Huang on 26/03/2024.
+//
+
+import Foundation
+import Abacus
+import Utilities
+
+final class TransferAnalytics {
+    func logDeposit(transferInput: TransferInput) {
+        log(event: .transferDeposit, transferInput: transferInput)
+    }
+
+    func logWithdrawal(transferInput: TransferInput) {
+        log(event: .transferWithdraw, transferInput: transferInput)
+    }
+
+    private func log(event: AnalyticsEvent, transferInput: TransferInput) {
+        let data = [
+            "chainId": Parser.standard.asString(transferInput.chainResource?.chainId),
+            "tokenAddress": transferInput.tokenResouce?.address,
+            "tokenSymbol": transferInput.tokenResouce?.symbol,
+            "slippage": Parser.standard.asString(transferInput.summary?.slippage),
+            "gasFee": Parser.standard.asString(transferInput.summary?.gasFee),
+            "bridgeFee": Parser.standard.asString(transferInput.summary?.bridgeFee),
+            "exchangeRate": Parser.standard.asString(transferInput.summary?.exchangeRate),
+            "estimatedRouteDuration": Parser.standard.asString(transferInput.summary?.estimatedRouteDuration),
+            "toAmount": Parser.standard.asString(transferInput.summary?.toAmount),
+            "toAmountMin": Parser.standard.asString(transferInput.summary?.toAmountMin)
```

Comment:
i forget, what happens if we send an optional instead of a primitive to firebase? Have you checked that it gets unwrapped before it reached FB/amplitude?

Reply:
There is a `.filterNils() as? [String: String]` call at the end, which ensures null values are removed.